### DISCUSSION TITLE
How to pass event handlers, like an onClick, as a prop in a reusable button component

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#63856] [Help]
### Summary

Hi,

I am attempting to create a very simple reusable button using NextJS. This component simply renders a button and props passed to it.

I have the following files:

**app/page.tsx**
```js
import { Button } from '@/components/button'

export default function Home() {
  return (
      <Button
        onClick={() => {
          console.log('FIRED')
        }}
      />
  )
}
```
**components/button.tsx**
```js
'use client'
import { ButtonHTMLAttributes } from 'react'

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement>

export function Button(props: ButtonProps) {
  return <button {...props}>Button</button>
}

```

This however, does not work. The following error is thrown:

```js
 ⨯ Error: Event handlers cannot be passed to Client Component props.
  <... onClick={function}>
               ^^^^^^^^^^
If you need interactivity, consider converting part of this to a Client Component.

```

I was wondering how am I supposed to fix this so that the button is reusable, allowing the onClick method to be a prop which could fire a different function depending on where it is used? It seems wrong to make the entire page a client component?

Thanks

**Comment by don-esteban** [comment]
You pass a function from a React Server Component (RSC) to a Client Component. Functions are not serializable and RSC/RCC run in different execution environments (Server/Browser).

Wrap it in a client component and i should work as expected.

**Reply by alex-r89** [reply]
Hi,

Thanks for the reply and the information. So how would I go about doing the part where you say "_Wrap it in a client component_"? 

As I said, I want the button to be reusable - so for example I may want to pass a different console log or function into the button to fire on the onClick, so simply moving the`() => { console.log('FIRED') }` into `button.tsx`  wont solve this problem, and I dont want the entire `page.tsx` file to be a client component just so that the above example onClick will work?

**Reply by don-esteban** [reply]
I‘m on a tablet and can‘t write much code. Your button is reusable but you can‘t pass a function from a RSC to a RCC. Create a new component, similar your page component and mark it with „client“. Use your button inside this component and pass it the desired function, which you also define inside this wrapper. Use this wrapper in your page component. I‘m not aware of a possibility passing functions between RSC and RCC. This would involve some serialization, deserialization und remote execution. Kind of Remote Procedure Call (RPC) I guess.

**Comment by don-esteban** [comment]
If you do this to examine/learn Next.js, it is just fine. Or do you have a real use case?  With and without a reusable button, you can‘t use an interactive button inside a RSC directly. If you have a real world use case, please describe it. Perhaps another approach can be found.
